src/disk_cache.rs

```
@@ -0,0 +1,1064 @@
+            .collect::<Vec<_>>();
+        let result_stream = stream::iter(futures).then(|fut| fut).boxed();
+
+        Ok(GetResult {
+            meta,
+            range,
+            attributes,
+            payload: GetResultPayload::Stream(result_stream),
+        })
+    }
+
+    // if an object is not cached before, maybe_prefetch_range will try to prefetch the object from the
+    // object store and save the parts into the local disk cache. the prefetching is helpful to reduce the
+    // number of GET requests to the object store, it'll try to aggregate the parts among the range into a
+    // single GET request, and save the related parts into local disks together.
+    // when it sends GET requests to the object store, the range is expected to be ALIGNED with the part
+    // size.
+    async fn maybe_prefetch_range(
+        &self,
+        location: &Path,
+        range: &Option<GetRange>,
+    ) -> object_store::Result<(ObjectMeta, Attributes)> {
+        let entry = self.storage.entry(location, self.part_size);
+        match entry.read_head().await {
+            Ok(Some((meta, attrs))) => return Ok((meta, attrs)),
+            Ok(None) => {}
+            Err(_) => {
+                // TODO: add a warning log
+            }
+        };
+
+        let aligned_opts = match &range {
+            None => GetOptions::default(),
+            Some(get_range) => GetOptions {
+                range: Some(self.align_get_range(get_range)),
+                ..Default::default()
+            },
+        };
+
+        let get_result = self.object_store.get_opts(location, aligned_opts).await?;
+        let result_meta = get_result.meta.clone();
+        let result_attrs = get_result.attributes.clone();
+        // swallow the error on saving to disk here (the disk might be already full), just fallback
+        // to the object store.
+        // TODO: add a warning log here.
+        self.save_result(get_result).await.ok();
```

ok, looks like it already ignores the error on saving to disk.